### ISSUE TITLE
Texplorer, PCC-Specific Properties and Thumbs

### ISSUE THREAD

**Issue by ghost** [ME3Explorer/ME3Explorer#322]
@KFreon -- Another change for the re-write :)

Texture properties are, in actuality, determined by PCCs. What I mean by this is the game reads the PCC, which points to the texture in a specific TFC, and from there the game streams the texture into the level (much simplified, I know). 

The fact that it works this way is why we can hijack this process and point PCCs to new textures files: CustTextures0.tfc, Textures_DLC_CON_BACKOFF, or whatever it might be. Furthermore, we can control exactly _which_ PCCs point to _which_ texture. The same texture can exist in 3 different TFCs in 3 different versions, and the texture that's loaded in game is ultimately controlled by the mount priority of the PCC. 

TL;DR, texture installation is PCC-specific and TFC-specific

What this means is that when it comes to modding, a "universal" Texture Properties pane as currently exists isn't ideal -- neither is a universal thumbnail for that matter:

![texp_pccselect](https://cloud.githubusercontent.com/assets/13645804/14581898/8a3b8fee-03ad-11e6-8854-8b92226aebb0.PNG)

The information we see in the above snapshot corresponds to vanilla or the most recent replacement. If the texture was replaced in all PCCs, then information is accurate. If the texture was replaced in only some PCCs, then it's not telling the whole story and can mislead the user.

---

What I'm suggesting is that the re-write (assuming it happens) might as well compensate for this. This is the reality of how users are using the toolset -- it's not the exception to the rule. I'm willing to help brainstorm  when the time comes, but for now, here are some thoughts on implementation (ME3/2-specific):
- When generating the tree, initial properties and thumbs are assigned. These are considered vanilla and they don't change. They are what's always loaded by default in the tool when opened.
- Upon selecting any texture, the tool has two new infoboxes, Texture File (TFC) & Level/Character File (PCC).
- The TFC box shows a list of all TFCs that contain the texture. Selecting different TFCs controls the appearance of the thumbnail, accurately showing the appearance of the texture in _that_ TFC.
- Once a TFC is selected, two other things happen. One, the Level File infobox is automatically populated with all the PCCs that point to that version of the texture. Two, the Properties box is also populated with the information that's specific to that version of the texture.
- This new TFC infobox could probably be used as an alternative to the "Change Destination TFC" button. The PCC infobox can be used in place of the existing one to control texture installation.

It would be complicated to implement, but it might actually work out pretty nicely. 

At any rate, some ideas. We can discuss more if/when you move forward with the re-write :)

**Comment by KFreon** [comment]
That all sounds fine, but as you mentioned, Texplorer doesn't control which TFC the game uses.
By that I mean, say you have the texture in 2 TFC's, the game is going to use the first PCC it comes across and whichever texture in whichever TFC that pcc points to.
So I'm curious as to what this multi TFC thing adds to the user?

**Comment by ghost** [comment]
> That all sounds fine, but as you mentioned, Texplorer doesn't control which TFC the game uses.
> By that I mean, say you have the texture in 2 TFC's, the game is going to use the first PCC it comes across and whichever texture in whichever TFC that pcc points to.

Well, not quite. It doesn't use the first one. The one used is controlled by mount priority. Whichever PCC has the highest mount priority is the one read by the game. So, whichever TFC _that_ PCC points to is what will be used.

> So I'm curious as to what this multi TFC thing adds to the user?

The multi-TFC thing is what will allow the user to see properties and thumbs specific to the texture as it exists in different TFCs. Right now, there's no way to access this information. The toolset shows the properties and thumb for the version of the texture that's been most recently replaced (or vanilla if never replaced). This doesn't account for if the texture exists as multiple versions. 

For example, any texture that's been modded exists in at least two versions (locations): Textures.tfc and CustTextures0.tfc. (I'm ignoring DLC mods and changing TFCs for this example). But, what if the user didn't install the new texture to every PCC? If they didn't, that means the game will use both versions of this texture. Yet, there's no way to see this information in Texplorer. The user will only ever see the new version of the texture as it exists in CustTextures0.tfc, Yet, the game is actually using two variants of this texture -- vanilla and modded. The user should be able to see thumbs and properties associated with each one.

If you throw a DLC mod into the mix, you might have 3 versions of the same texture in 3 TFCs. The user should be able to see what the texture looks like in each TFC and what the associated properties are. From there,  "Change destination TFC" comes into play.

**Comment by KFreon** [comment]
But...I thought as you said the mount priority was the important part.
So it doesn't matter if they install it to different PCC's, one of them is going to get overridden all the time.
Like if you had a DLC, doesn't it override all the textures it contains?
Or haven't I had enough coffee this morning?

**Comment by ghost** [comment]
I think (maybe, lol!), you're still missing the part PCC-specific texture installation. About how when a texture is installed, that installation may not affect all PCCs, which means both the new AND old versions of the texture are used in the game.

Let me try once more. If it still isn't making sense we can voice chat on Skype, whenever :)

We'll use Ashley's face DIFF as an example. This texture is referenced by 18 PCCs in my install: 14 in base game, 2 in From Ashes, and 2 in Collecters Edition. These are all _unique_ PCCs.

Let's say I want to install a new texture for her face, but I only want it to affect CE content. So, I disable all but 2 of the PCCs that are specific to the Online Pass DLC: BioH_Ashley_02.pcc and BIoH_Ashley_02_Explore.pcc. 

Then, I install the new textures. This means that now those two PCCs use a new DIFF -- the modded one, and they point to CustTextures0.tfc. The Texplorer thumbnail will reflect this change, as will the Properties. This data is _static_ it will not change until if/when I change the texture again.

But, what about the other 16 PCCs that reference Ashley's facial DIFF? They still use the vanilla version in the CharTextures.tfc. Yet, since I've replaced the texture, there's no way for me to go back later and check in Texplorer as to what the properties are for the texture in CharTextures.tfc, or what the thumb for that version of the texture looks like. And, since the properties/thumb info doesn't change when I select/deselect different PCCs associated with the DIFF, I actually **don't know** which ones are using the modded texture versus the vanilla one.

I could now bring a DLC mod into the picture, but that will complicate things further. Does the above  make sense now?

**Comment by KFreon** [comment]
Yes but I thought when anything from one of the DLC's was loaded, that DLC PCC would override the basegame one right?
So sometimes when you specify PCC's to update thinking it will only affect the area the pcc is usually seen in, that might not be the case since it's been overridden by a DLC one.

Right?

**Comment by ghost** [comment]
@KFreon -- Overriding PCCs must have the same name. That's how it works in single-player. Most PCCs in DLC are unique, but there are some cases of overriding. Generally, these are Normandy levels.

All character-specific content in DLCs for alternate appearances are separate files; these _do not_ override vanilla content in a strict sense. They supplement it. And, in fact, vanilla content can still sometimes (sort of) override them.

Here's an example, just with hair meshes. Last night during testing I installed a different hair mesh to Ashley's 2 CE PCCs. After loading a save, selecting her in the squad select area, and changing her into the CE outfit, her hair correctly used the modded hair style on the talent assignment screen. However, once the mission started, her character reverted to the vanilla hairstyle -- despite being in the CE "appearance". This was unexpected (it means as far as her hair mesh is concerned the game is preferentially loading from a different file), but regardless, I have effectively created a situation where the game uses two hairstyles for Ashley. The same thing can be done for textures.

From your response, I still don't think we're understanding each other. Even with DLC out of the picture and installing only to the base game, this is still an issue. Ultimately, it's is due to the ability to control PCC selection during the texture install process (which is something we definitely want to retain).

Let's just have a chat on Skype when we have time. It's a complex situation, and hard to discuss adequately via written word :)
